### PR TITLE
[Storage] [DataMovement] Fix for skipped and failed transfer status and Single Upload / Download transfers

### PR DESCRIPTION
**Pull request by amnguye**
tip: ignore the json files.

Found an issue with overwrite enabled not passing back correct transfer status, so included it in this PR.

Also added in the transfer options for upload to guarantee the single put blob upload call, because before we were defaulting to whatever the BlobClient had before.

I was going to add more test recordings in this PR, but making and pushing commits got slow.

### CODE REVIEW

File: sdk/storage/Azure.Storage.DataMovement/src/Models/SingleTransferOptions.cs [hunk 1]
```diff
@@ -106,5 +103,13 @@ public bool Equals(SingleTransferOptions obj)
         public event SyncAsyncEventHandler<TransferFailedEventArgs> TransferFailed;
 
         internal SyncAsyncEventHandler<TransferFailedEventArgs> GetFailed() => TransferFailed;
+
+        /// <summary>
+        /// Number of single transfers skipped during Transfer due to no overwrite allowed as specified in
+        /// <see cref="CreateMode"/>
```
Comment (jalauzon-msft) [inline sdk/storage/Azure.Storage.DataMovement/src/Models/SingleTransferOptions.cs:109]: nit: Maybe this comment should read more like the one above? It seems like this isn't really the number of skipped transfers but rather a collection of those skipped transfers?

File: sdk/storage/Azure.Storage.DataMovement/src/Shared/DataTransfer.cs [hunk 2]
```diff
@@ -77,7 +82,7 @@ private static void VerifyTaskCompleted(bool isCompleted)
             {
                 if (Debugger.IsAttached)
                 {
-                    Debugger.Break();
+                    //Debugger.Break();
```
Comment (jalauzon-msft) [inline sdk/storage/Azure.Storage.DataMovement/src/Shared/DataTransfer.cs:85]: Was this intentionally commented out?

File: sdk/storage/Azure.Storage.DataMovement/src/TransferJobInternal.cs [hunk 3]
```diff
@@ -123,6 +123,18 @@ internal abstract class TransferJobInternal : IDisposable
         /// </summary>
         public SyncAsyncEventHandler<TransferFailedEventArgs> TransferFailedEventHandler { get; internal set; }
 
+        /// <summary>
+        /// Number of single transfers skipped during Transfer due to no overwrite allowed as specified in
+        /// <see cref="StorageResourceCreateMode.Skip"/>
```
Comment (jalauzon-msft) [inline sdk/storage/Azure.Storage.DataMovement/src/TransferJobInternal.cs:128]: Same comment as above on comment. Super nit though

File: sdk/storage/Azure.Storage.DataMovement/tests/Shared/DataMovementBlobTestBase.cs [hunk 4]
```diff
@@ -579,5 +580,89 @@ internal static async Task DownloadAndAssertAsync(Stream stream, BlobBaseClient
                     actual.AsSpan(0, count).ToArray());
             }
         }
+
+        internal async Task<AppendBlobClient> CreateAppendBlob(
+    BlobContainerClient containerClient,
+    string localSourceFile,
+    string blobName,
+    long size)
```
Comment (jalauzon-msft) [inline sdk/storage/Azure.Storage.DataMovement/tests/Shared/DataMovementBlobTestBase.cs:588]: nit: indents

**Review by jalauzon-msft — APPROVED** [review]

File: sdk/storage/Azure.Storage.DataMovement/src/Shared/DataTransfer.cs (hunk 2 above)
Comment (amnguye) [reply sdk/storage/Azure.Storage.DataMovement/src/Shared/DataTransfer.cs:85]: unintentionally good catch.

File: eng/nunit.runsettings [hunk 5]
```diff
@@ -22,7 +22,7 @@
 		<!-- <Parameter name="TestMode" value="Record" /> -->
 
 		<!-- Enable running Fiddler -->
-		<!-- <Parameter name="EnableFiddler" value="true" /> -->
+		<!--  <Parameter name="EnableFiddler" value="true" /> -->
```
Comment (amnguye) [inline eng/nunit.runsettings:25]: Remove this random space I accidentally did.

**Review by jalauzon-msft — APPROVED** [review]

File: sdk/storage/Azure.Storage.DataMovement/src/CommitChunkHandler.cs [hunk 6]
```diff
@@ -70,7 +83,21 @@ public struct Behaviors
...
+            // and that's in the worst case that we never read from the channel and had a maximum chunk blob.
+            _stageChunkChannel = Channel.CreateUnbounded<StageChunkEventArgs>(
+                new UnboundedChannelOptions()
+                {
+                    // Single reader is required as we can only read and write to bytesTransferred value
+                    SingleReader = true,
+                });
+            _cancellationTokenSource = new CancellationTokenSource();
+            _processStageChunkEvents = Task.Run(() => NotifyOfPendingStageChunkEvents());
+
```
Comment (amnguye) [inline sdk/storage/Azure.Storage.DataMovement/src/CommitChunkHandler.cs:98]: Will cherry pick these changes out and put in a different PR for future review and testing.

https://github.com/Azure/azure-sdk-for-net/issues/32854

File: sdk/storage/Azure.Storage.DataMovement/src/CommitChunkHandler.cs [hunk 7]
```diff
@@ -82,12 +109,25 @@ public struct Behaviors
...
+            {
+                _cancellationTokenSource.Cancel();
+            }
+            _cancellationTokenSource.Dispose();
+
+            if (_currentBytesSemaphore != default)
+            {
+                _currentBytesSemaphore.Dispose();
+            }
+            DipsoseHandlers();
```
Comment (amnguye) [inline sdk/storage/Azure.Storage.DataMovement/src/CommitChunkHandler.cs:127]: Will cherry pick these changes out and put in a different PR for future review and testing.

https://github.com/Azure/azure-sdk-for-net/issues/32854

File: sdk/storage/Azure.Storage.DataMovement/src/CommitChunkHandler.cs [hunk 8]
```diff
@@ -100,28 +140,61 @@ private async Task CommitBlockEvent(StageChunkEventArgs args)
...
+                await _stageChunkChannel.Writer.WriteAsync(args, _cancellationToken).ConfigureAwait(false);
             }
             else
             {
-                // Set status to completed
                 await _invokeFailedEventHandler(new Exception("Failure on Stage Block")).ConfigureAwait(false);
             }
         }
 
+        private async Task NotifyOfPendingStageChunkEvents()
```
Comment (amnguye) [inline sdk/storage/Azure.Storage.DataMovement/src/CommitChunkHandler.cs:152]: Will cherry pick these changes out and put in a different PR for future review and testing.

https://github.com/Azure/azure-sdk-for-net/issues/32854

File: sdk/storage/Azure.Storage.DataMovement/src/DownloadChunkHandler.cs [hunk 9]
```diff
@@ -78,18 +90,29 @@ public struct Behaviors
...
+                {
+                    // Single reader is required as we can only read and write to bytesTransferred value
+                    SingleReader = true,
+                });
+            _channelCancellationSource = new CancellationTokenSource();
+            _processDownloadRangeEvents = Task.Run(() => NotifyOfPendingChunkDownloadEvents());
+
+            _expectedLength = expectedLength;
+            _ranges = ranges;
+
```
Comment (amnguye) [inline sdk/storage/Azure.Storage.DataMovement/src/DownloadChunkHandler.cs:115]: Will cherry pick these changes out and put in a different PR for future review and testing.

https://github.com/Azure/azure-sdk-for-net/issues/32854

File: sdk/storage/Azure.Storage.DataMovement/src/DownloadChunkHandler.cs [hunk 10]
```diff
@@ -121,75 +142,125 @@ public struct Behaviors
...
+            else
+            {
+                // Report back failed event.
+                await InvokeFailedEvent(new Exception("Unexpected error: Experienced failed download range argument. " +
+                    $"Range: {args.Offset} - {args.BytesTransferred} with Transfer ID: {args.TransferId}")).ConfigureAwait(false);
+            }
+        }
+
+        private async Task NotifyOfPendingChunkDownloadEvents()
+        {
```
Comment (amnguye) [inline sdk/storage/Azure.Storage.DataMovement/src/DownloadChunkHandler.cs:183]: Will cherry pick these changes out and put in a different PR for future review and testing.

https://github.com/Azure/azure-sdk-for-net/issues/32854

**Pull request merged**